hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskControllerImpl.java

```diff
@@ -416,19 +459,358 @@ private void processEventBatchDirectly(OMUpdateEventBatch events) {
+    }
+    return null;
+  }
+
+  private ReconTaskController.ReInitializationResult validateRetryCountAndDelay() {
+    // Check if we should retry based on timing for iteration-based retries
+    long currentTime = System.currentTimeMillis();
+    if (eventProcessRetryCount.get() > 0) {
+      // Check if 2 seconds have passed since last iteration
+      long timeSinceLastRetry = currentTime - lastRetryTimestamp.get();
+      if (timeSinceLastRetry < RETRY_DELAY_MS) {
+        LOG.debug("Skipping retry, only {}ms since last retry attempt (need {}ms)",
+            timeSinceLastRetry, RETRY_DELAY_MS);
+        return ReInitializationResult.RETRY_LATER;
+      }
+      LOG.info("Attempting retry (retry attempt count: {}, delay: {}ms)",
+          eventProcessRetryCount.get() + 1, timeSinceLastRetry);
+    }
+    return getEventRetryResult();
+  }
+
+  /**
+   * Handle iteration failure by updating retry counters.
+   */
+  private void handleEventFailure() {
+    long currentTime = System.currentTimeMillis();
+    lastRetryTimestamp.set(currentTime);
+    eventProcessRetryCount.getAndIncrement();
+    LOG.error("Event processing failed {} times.", eventProcessRetryCount);
+  }
+  
+  /**
+   * Determine the appropriate retry result based on current event retry count.
+   */
+  private ReconTaskController.ReInitializationResult getEventRetryResult() {
+    if (eventProcessRetryCount.get() >= MAX_EVENT_PROCESS_RETRIES) {
+      LOG.warn("Maximum iteration retries ({}) exceeded, resetting counters and signaling full snapshot fallback",
+          MAX_EVENT_PROCESS_RETRIES);
+      resetRetryCounters();
+      return ReconTaskController.ReInitializationResult.MAX_RETRIES_EXCEEDED;
+    }
+    return null;
+  }
+
+  public void drainEventBufferAndCleanExistingCheckpoints() {
+    // First drain all events to check for any ReconTaskReInitializationEvent that need checkpoint cleanup
+    List<ReconEvent> drainedEvents = new ArrayList<>();
+    int drainedCount = eventBuffer.drainTo(drainedEvents);
+    
+    if (drainedCount > 0) {
+      LOG.info("Drained {} events from buffer before clearing. Checking for checkpoint cleanup.", drainedCount);
+      
+      // Check for any ReconTaskReInitializationEvent and cleanup their checkpoints
+      for (ReconEvent event : drainedEvents) {
+        if (event instanceof ReconTaskReInitializationEvent) {
+          ReconTaskReInitializationEvent reinitEvent = (ReconTaskReInitializationEvent) event;
+          ReconOMMetadataManager checkpointedManager = reinitEvent.getCheckpointedOMMetadataManager();
+          if (checkpointedManager != null) {
+            LOG.info("Cleaning up unprocessed checkpoint from drained ReconTaskReInitializationEvent");
+            cleanupCheckpoint(checkpointedManager);
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public void updateOMMetadataManager(ReconOMMetadataManager omMetadataManager) {
+    LOG.info("Updating OM metadata manager");
```

Comment:
there will be frequent info log as its called every time in while loop for delta update